driver/reducecalculation_driver.hpp

```diff
@@ -237,27 +312,51 @@ int ReduceCalculationDriver<Tgpu, Tref>::AllocateBuffersAndCopy()
     uint32_t ctx = 0;
 
     in_dev        = std::unique_ptr<GPUMem>(new GPUMem(ctx, in_sz, sizeof(Tgpu)));
-    out_dev       = std::unique_ptr<GPUMem>(new GPUMem(ctx, out_sz, sizeof(Tgpu)));
+    out_dev       = IsLogicalCalculation() ? std::make_unique<GPUMem>(ctx, out_sz, sizeof(bool))
+                                           : std::make_unique<GPUMem>(ctx, out_sz, sizeof(Tgpu));
     workspace_dev = std::unique_ptr<GPUMem>(new GPUMem(ctx, ws_sizeInBytes, sizeof(std::byte)));
 
-    in      = std::vector<Tgpu>(in_sz, static_cast<Tgpu>(0));
-    out     = std::vector<Tgpu>(out_sz, static_cast<Tgpu>(0));
-    outhost = std::vector<Tref>(out_sz, static_cast<Tref>(0));
+    in = std::vector<Tgpu>(in_sz, static_cast<Tgpu>(0));
 
-    for(int i = 0; i < in_sz; ++i)
+    if(IsLogicalCalculation())
     {
-        in[i] = prng::gen_A_to_B<Tgpu>(static_cast<Tgpu>(0.0), static_cast<Tgpu>(1.0));
+        logical_out     = std::vector<uint8_t>(out_sz, 0);
+        logical_outhost = std::vector<uint8_t>(out_sz, 0);
+    }
+    else
+    {
+        out     = std::vector<Tgpu>(out_sz, static_cast<Tgpu>(0));
+        outhost = std::vector<Tref>(out_sz, static_cast<Tref>(0));
+    }
```

Comment:
why using `if-else` at here, the code seem duplicate

Reply:
because there're 2 types of outputs: `logical_out` and normal `output`. I dont merge those 2 outputs because I want to explicitly declare `logical_out` as type `uint8_t`. By using this, I dont have to pass `output_type` in template Driver. Of course I can pass `output_type` in template class Driver, this is just about coding style.

Reply:
I see that passing `output_type` in template Driver is cleaner and more abstracted than separating between `logical_out` and normal `output`